### PR TITLE
[opentitantool] Consolidate GPIO declarations before applying

### PR DESCRIPTION
**Pull request by jesultra**
Sometimes a pin may appear under several levels of aliases in multiple configuration files, and one file may specify e.g. OpenDrain mode without specifying a level, while anoter file specifies a default "high" level, without mentioning the mode.  If such two declarations used different names for the same pin, the existing code could end up applying the level before the mode, which HyperDebug may object to.  (As it rejects setting level on pins in input mode.)

This change makes it such that before applying any pin configuration, all pin configurations ultimately referring to the same pin will be "consolidated", such that mode, level and weak pull will all be applied atomically.  This both prevents glitches on the signals, and prevents the aforementioned HyperDebug issue.

Some other undesirable behavior is fixed along the way, such as previously if two gpio entries in the same file mentioned the same pin name, but with different configuration, the latter would silently override the former.  Doing the same now will cause an error (unless the configurations are "compatible").

Signed-off-by: Jes B. Klinke <jbk@chromium.org>
Change-Id: I896033cf9d3c13dff0157164c4a7bf62a38657c5

### CODE REVIEW

File: sw/host/opentitanlib/src/app/mod.rs [hunk 1]
```diff
@@ -56,19 +91,117 @@ pub struct TransportWrapper {
...
+        }
+        if let Some(level) = pin_conf.level {
+            conf_entry.level = Some(level);
+        }
+        pin_conf_list.push((pin_conf.name.to_string(), conf_entry))
+    }
+
+    pub fn add_configuration_file(&mut self, file: config::ConfigurationFile) -> Result<()> {
+        // Merge content of configuration file into pin_map and other
+        // members.
```
Comment (marnovandermaas) [inline sw/host/opentitanlib/src/app/mod.rs:129]: Nit: this can be on the same line.

File: sw/host/opentitanlib/src/app/mod.rs [hunk 2]
```diff
@@ -272,3 +321,44 @@ impl TransportWrapper {
...
+        Some(v) => {
+            if v.eq(&name) {
+                name
+            } else {
+                map_name(map, v)
+            }
+        }
+        None => name,
+    }
+}
```
Comment (marnovandermaas) [inline sw/host/opentitanlib/src/app/mod.rs:340]: This looks like the same function as above? It would have been easier to review if moving functions around was in a separate commit.

**Review by marnovandermaas — APPROVED** [review]
I'm not a huge rust expert, but the code looks fine to me. I just added a few nits below.

File: sw/host/opentitanlib/src/app/mod.rs [hunk 3]
```diff
@@ -272,3 +321,44 @@ impl TransportWrapper {
...
+    }
+
+    fn set_mode(&self, _mode: PinMode) -> Result<()> {
+        Ok(())
+    }
+
+    fn set_pull_mode(&self, _mode: PullMode) -> Result<()> {
+        Ok(())
+    }
+}
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/app/mod.rs:364]: Should we consider using a NullPin to be at least a warnable problem? (ie: add `log::warn!` statements to these functions)?

**Review by cfrantz — APPROVED** [review]

File: sw/host/opentitanlib/src/app/mod.rs (hunk 2 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/app/mod.rs:340]: Yes, my problem, however, is that I am separating a struct into two structs, with some member function going into one and some into the other, and a single one `map_name` needing to be accessed from both, which is why it now sits outside both.  It is quite hard to separate the textual moving around of functions, from the semantic breaking of the struct into two.

File: sw/host/opentitanlib/src/app/mod.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/app/mod.rs:129]: Thanks, done.

File: sw/host/opentitanlib/src/app/mod.rs (hunk 3 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/app/mod.rs:364]: That is a good suggestion, done.

**Comment by jesultra** [comment]
This time CI gave a different error, something about the operation "being cancelled".  Closing and reopening once again, in an attempt to get a successful run.
